File: src/main/java/com/dambra/paul/moduluschecker/chain/checks/ModulusTotal.java
package com.dambra.paul.moduluschecker.chain.checks;

import com.dambra.paul.moduluschecker.Account.BankAccount;
import com.dambra.paul.moduluschecker.As;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Streams;

final class ModulusTotal {

    static int calculate(BankAccount account, ImmutableList<Integer> weights) {
        return Streams
            .zip(
                account.allDigits(),
                weights.stream(),
                (l, r) -> l * r
            )
            .reduce(0, Integer::sum);
    }

    static int calculateDoubleAlternate(BankAccount account, ImmutableList<Integer> weights) {
        return Streams
            .zip(
                account.allDigits(),
                weights.stream(),
                (l, r) -> l * r
            )
            .map(String::valueOf)
            .flatMap(As::integerStream)
            .reduce(0, Integer::sum);
    }
}

File: src/main/java/com/github/pauldambra/moduluschecker/chain/checks/ModulusTotal.kt
package com.github.pauldambra.moduluschecker.chain.checks

import com.github.pauldambra.moduluschecker.As
import com.github.pauldambra.moduluschecker.account.BankAccount
import com.google.common.collect.Streams

internal object ModulusTotal {

    fun calculate(account: BankAccount, weights: List<Int>): Int {
        return Streams
          .zip(
            account.allDigits(),
            weights.stream()
          ) { l, r -> l!! * r!! }
          .reduce(0) { a, b -> Integer.sum(a, b) }
    }

    fun calculateDoubleAlternate(account: BankAccount, weights: List<Int>): Int {
        return Streams
          .zip(
            account.allDigits(),
            weights.stream()
          ) { l, r -> l!! * r!! }
          .map<String> { it.toString() }
          .flatMap<Int> { As.integerStream(it) }
          .reduce(0) { a, b -> Integer.sum(a, b) }
    }
}
